quill-sql/src/main/scala/io/getquill/context/sql/Prepare.scala

```diff
@@ -32,21 +26,20 @@ object Prepare {
           other.show
       }
 
-    val (entity, insert) = ExtractEntityAndInsertAction(bindedAst)
-    val isInsert = insert.isDefined
-
-    val generated = if (isInsert) entity.flatMap(nameGeneratedColumn) else None
-
-    (sqlString, idents, generated)
+    (sqlString, idents, returning.map(prop => returningColumn(ast, prop)))
   }
 
-  private def nameGeneratedColumn(entity: Entity)(implicit n: NamingStrategy): Option[String] = {
+  private def returningColumn(ast: Ast, prop: String)(implicit n: NamingStrategy) = {
+    val entity = CollectAst(ast) {
```

Comment:
But a query (or an AST) with two Insert elements is an invalid query, and the `QueryDsl` does not allow using `insert` more than one time. I think we are safe here.